keps/sig-node/3386-kubelet-evented-pleg/README.md

```diff
@@ -180,35 +234,33 @@ enum ContainerEventType {
     CONTAINER_DELETED_EVENT = 3;
 }
 ```
-### Events Filter
-Events can be filtered to retrieve only subset of events,
+### Pod Status Update in the Cache
 
-```protobuf=
-message GetEventsRequest {
-    // Optional to filter a list of events.
-    GetEventsFilter filter = 1;
-}
-```
+While using `Evented PLEG`, the existing `Generic PLEG` is set to relist with the increased period. But in case `Evented PLEG` faces temporary disruptions in the grpc connection with the runtime, there is a chance that when the normalcy is restored the incoming buffered events (which are outdated now) might end up overwriting the latest pod status in the cache updated by the `Generic PLEG`. Having a cache setter that only updates if the pod status in the cache is older than the current pod status helps in mitigating this issue.
 
-```protobuf=
-// GetEventsFilter is used to filter a list of events.
-// All those fields are combined with 'AND'
-message GetEventsFilter {
-    // ID of the container, sandbox.
-    string id = 1;
-    // LabelSelector to select matches.
-    // Only api.MatchLabels is supported for now and the requirements
-    // are ANDed. MatchExpressions is not supported yet.
-    map<string, string> label_selector = 2;
-}
-```
+At present kubelet updates the cache using the [Set function](https://github.com/kubernetes/kubernetes/blob/7f129f1c9af62cc3cd4f6b754dacdf5932f39d5c/pkg/kubelet/container/cache.go#L101).
 
-### Kubelet Changes
+Pod status should be updated in the cache only if the new status update has timestamp newer than the timestamp of the already present in the cache.
 
-Kubelet generates [PodLifecycleEvent](https://github.com/kubernetes/kubernetes/blob/release-1.24/pkg/kubelet/pleg/pleg.go#L41) using [relisting](https://github.com/kubernetes/kubernetes/blob/050f930f8968874855eb215f0c0f0877bcdaa0e8/pkg/kubelet/pleg/generic.go#L150). These `PodLifecycleEvents` get [used](https://github.com/kubernetes/kubernetes/blob/050f930f8968874855eb215f0c0f0877bcdaa0e8/pkg/kubelet/kubelet.go#L2060) in kubelet's sync loop to infer the state of the container. e.g. to determine if the [container has died](https://github.com/kubernetes/kubernetes/blob/050f930f8968874855eb215f0c0f0877bcdaa0e8/pkg/kubelet/kubelet.go#L2118).
+![Modified Cache Setter](./modified-cache-setter.png)
+
+```go
+func (c *cache) Set(id types.UID, status *PodStatus, err error, timestamp time.Time) (updated bool) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	// Set the value in the cache only if it's not present already
+	// or the timestamp in the cache is older than the current update timestamp
+	if val, ok := c.pods[id]; !ok || val.modified.Before(timestamp) {
+		c.pods[id] = &data{status: status, err: err, modified: timestamp}
+		c.notify(id, timestamp)
+		return true
+	}
+	return false
```

Comment:
probably some implementation details and we dont need to address in KEP, just curious if `Set` returns a bool, then it means the two PLEGs will behave differently based on whether the `Set` succeeded right?

Reply:
Also, this change also effectively deprecates the global timestamp used by the pod cache, because every cache update is worth calling `notify`, right?

Reply:
In Generic PLEG, the [updateCache](https://github.com/kubernetes/kubernetes/blob/e9f28c02543fd5601bd2d5b6b8488afa0a550c5f/pkg/kubelet/pleg/generic.go#L250) function which calls that Set can indicate if the Set was successful or not. Then when it comes to [sending PodLifeCycleEvent for that pod](https://github.com/kubernetes/kubernetes/blob/e9f28c02543fd5601bd2d5b6b8488afa0a550c5f/pkg/kubelet/pleg/generic.go#L277), we can skip that pod. 

Similarly in Evented PLEG, if the Set doesn't update the cache then we will skip sending the PodLifeCycleEvent for that pod. 

This way both the PLEGs will behave same. But you are right, it's an implementation detail that we can handle it in actual code.